pkg/build/registry/buildlog/rest.go

```diff
@@ -119,21 +123,190 @@ func (r *REST) Get(ctx apirequest.Context, name string, opts runtime.Object) (ru
+					glog.Errorf("error: failed to stream logs for build pod: %s/%s container: %s, due to: %v", build.Namespace, buildPodName, status.Name, err)
+					return
+				}
+
+				// if we successfully streamed anything, don't wait before the next iteration
+				// of init container checking/streaming.
+				sleep = false
+
+				// we are done with this container, we can ignore on future iterations.
+				doneWithContainer[status.Name] = true
+
+				// no point in processing more init containers once we've seen one that failed,
+				// no additional init containers will run.
+				if initFailed {
+					break
+				}
+			} // loop over all the initcontainers
+
+			// if we're not in log follow mode, don't keep waiting on container logs once
+			// we've iterated all the init containers once.
+			if !buildLogOpts.Follow {
+				break
+			}
+			// don't iterate too quickly waiting for the next init container to run unless
+			// we did some log streaming during this iteration.
+			if sleep {
+				time.Sleep(time.Second)
+			}
+
+			// loop over the pod until we've seen all the initcontainers enter the running state and
+		} // streamed their logs, or seen a failed initcontainer, or we weren't in follow mode.
+
+		// done handling init container logs, get the main container logs, unless
+		// an init container failed, in which case there will be no main container logs and we're done.
+		if !initFailed {
+
+			// Wait for the main container to be running, this can take a second after the initcontainers
+			// finish so we have to poll.
+			err := wait.PollImmediate(time.Second, 10*time.Minute, func() (bool, error) {
+				obj, err = r.PodGetter.Get(ctx, buildPodName, &metav1.GetOptions{})
+				if err != nil {
+					s := fmt.Sprintf("error while getting build logs, could not retrieve build pod %s/%s : %v", build.Namespace, buildPodName, err.Error())
+					pipeStreamer.In.Write([]byte(s))
+					return false, err
+				}
+				buildPod = obj.(*kapi.Pod)
+				// we can get logs from a pod in any state other than pending.
+				if buildPod.Status.Phase != kapi.PodPending {
+					return true, nil
+				}
+				return false, nil
+			})
+			if err != nil {
+				glog.Errorf("error: failed to stream logs for build pod: %s/%s due to: %v", build.Namespace, buildPodName, err)
+				return
+			}
+
+			containerLogOpts := buildapi.BuildToPodLogOptions(buildLogOpts)
+			containerLogOpts.Container = ""
+			// never follow logs for terminated pods, it just causes latency in streaming the result.
+			if buildPod.Status.Phase == kapi.PodFailed || buildPod.Status.Phase == kapi.PodSucceeded {
+				containerLogOpts.Follow = false
+			}
+
+			if err := r.pipeLogs(ctx, build.Namespace, buildPodName, containerLogOpts, pipeStreamer.In); err != nil {
+				glog.Errorf("error: failed to stream logs for build pod: %s/%s due to: %v", build.Namespace, buildPodName, err)
+				return
+			}
+		}
+	}()
+
+	return &pipeStreamer, nil
```

Comment:
Does it make sense to break up this big brand new chunk of code into some private method/funcs?